### PR TITLE
linkcheck builder: begin using session-based HTTP requests

### LINKED ISSUE #11324: linkcheck builder: begin using requests.Session functionality during linkchecking
**Is your feature request related to a problem? Please describe.**
At the moment, the `linkcheck` builder performs individual `request.get` (or similar HTTP request method) operations during linkchecking, without any explicit connection or session pooling.

This may be inefficient, because it seems likely that for many use cases, linkchecking will make multiple requests to the same host (because documentation references are likely to have host-locality).

**Describe the solution you'd like**
Confirmation that connection pooling is not currently in use would be a good starting point; in other words: we should confirm that linkchecking of multiple URLs on a single host results in multiple TCP connections.  Ideally this should be written as a test case.

If we can confirm that the problem exists, then we may be able to use some of the [`Session` object functionality](https://requests.readthedocs.io/en/latest/user/advanced/#session-objects) from the `requests` library that's already in use here to enable connection pooling.

**Describe alternatives you've considered**
None so far, although open to suggestions (and improvements on the definition of this feature request).

**Additional context**
- See note / suggestion about use of context managers: https://github.com/sphinx-doc/sphinx/issues/11317#issuecomment-1508319197

### PR DESCRIPTION
**Pull request by jayaddison**
### Feature or Bugfix
- Bugfix
- Refactoring

### Purpose
- Reduce the number of HTTP connections used during linkchecking, reducing network resource usage.

### Detail
- Resolves #11324.

### Relates
- Re-spin of #11340 following some interim refactors and cleanups of the `linkcheck` builder.

### CODE REVIEW

File: sphinx/util/requests.py [hunk 1]
```diff
@@ -77,3 +77,38 @@ def head(url: str,
...
+
+
+class _Session(requests.Session):
+
+    def get(self, url: str, **kwargs: Any) -> requests.Response:  # type: ignore[override]
+        """Sends a GET request like requests.get().
+
+        This sets up User-Agent header and TLS verification automatically."""
+        headers = kwargs.setdefault('headers', {})
+        headers.setdefault('User-Agent', kwargs.pop('_user_agent', _USER_AGENT))
```
Comment (jayaddison) [inline sphinx/util/requests.py]: The `kwargs.pop` pattern used here - and also for `_tls_info` - might not be ideal.  I began using that due to line length limit complaints by `flake8` with the additional indenting required for class methods compared to a top-level module method.

**Comment by AA-Turner** [comment]
Should we rewrite the existing get and head in terms of a session? This is how ``requests`` implements its module-level request methods, might make sense to take inspiration from that.

A

**Comment by jayaddison** [comment]
It's _probably_ a good idea to re-use the same code paths for both methods.. my rationale when duplicating the implementations was to avoid modifying any non-linkcheck-builder code.

I'll take another look into consolidating them into shared implementations.

**Comment by jayaddison** [comment]
What do you think of the approach in b7d7fcfc9ba65680cb41e6366fe8da34b9f7cd3a ?

File: sphinx/util/requests.py (hunk 1 above)
Comment (jayaddison) [reply sphinx/util/requests.py]: I think I should probably take one more attempt at restoring the type-checking for `_tls_info` and `_user_agent`.  One of those things that could exist for a long time if it isnt'f fixed during this changeset.

Comment (jayaddison) [reply sphinx/util/requests.py]: (done)

Comment (jayaddison) [reply sphinx/util/requests.py]: Ah.. but actually using those parameters would make sense!  Oops.  One moment..

Comment (jayaddison) [reply sphinx/util/requests.py]: OK.  I made a bit of a mess of that, but I think that both of the session-based methods should now be reading the `_user_agent` and `_tls_info` arguments much like the previous, individual-request-based methods were, and with the same level of argument typechecking (at least when used through the `_Session` class).

**Comment by AA-Turner** [comment]
I've pushed a commit to only implement the lower-level ``Session.request``, to avoid duplication -- hopefully all looks good!

A

**Comment by jayaddison** [comment]
> I've pushed a commit to only implement the lower-level `Session.request`, to avoid duplication -- hopefully all looks good!

Nice - yep, that looks much better to me :+1:

**Comment by AA-Turner** [comment]
Thanks @jayaddison -- this is the culmination of quite a bit of preparatory work, thank you for driving this all through!

A

**Comment by jayaddison** [comment]
Thank you very much @AA-Turner :)

You're welcome.  I'll follow along with the next release to listen for user feedback.

**Pull request merged**